### DISCUSSION TITLE
flat config in ide support and monorepo

### DISCUSSION THREAD

**Discussion by skimhugo** [eslint/eslint#16202] [General]
I think another import issue about flat config is ide support and monorepo like pnpm.  And ide support in a monorepo env.

**Comment by nzakas** [comment]
Do you have a question?

**Comment by sam3k** [comment]
Hi @skimhugo, can you elaborate on your thoughts here? What are the other import issues? Can you link to these issues and also provide more detail information?

**Reply by airtonix** [reply]
> What are the other import issues?

@skimhugo probably meant to say "important".

And thus having brought out my speakwrite machine, I'll gleefully appropriate and hijack the question :> 

In a monorepo (a good typical example is one created by nx.dev), you have one main `.eslintrc.json` in the root, and then throughout the monorepo each package locally defines their customisations.

You guys have spent a lot of time writing **a lot** of words to tell us all that this was Bad™️. 
Although I can't really tell because my eyes glazed over as my peasant mouth breathing brain struggled to read all the incredibly large and well educated words used.

I personally liked the proximity of eslint overrides to the package requiring them.

It seems like eslint now only looks for one config file and any proximity would mean something like: 

```ts
import { EslintDsButtonConfig } from '@ourmonorepo/ds-button';
import { EslintDsBoxConfig } from '@ourmonorepo/ds-box';
import { EslintDsFormsConfig } from '@ourmonorepo/ds-forms';
// over 9000 other packages

export default [
  {
    files: ['pkg/ds/button/**/*.tsx'],
    ...EslintDsButtonConfig
  },
  // over 9000 other similar things,
  {
    files: ['**/*'],
    ...the global defaults.
  }
]
```


Am I wrong? 

Have you created a simplier intro to all this amazingly wonderful complexity for simple smooth brain mouse using peasants like myself that can only read single syallable words?

❤️

**Reply by nzakas** [reply]
I appreciate the feedback, but perhaps we could do it with a little less sarcasm? :)

Yes, ESLint now only looks for one config file, starting in the current working directory and working its way up the filesystem. 

However, you can also override the file ESLint uses by passing the `-c` command line argument:
https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file-resolution

This should give plenty of customization opportunities for monorepos.

**Reply by skimhugo** [reply]
There are some interesting rules regarding eslint and pnpm, such as pnpm's default hosting of eslint and eslint plugins in the root workspace. Nowadays, monorepo seems to be everyone's choice. However, combining monorepo with an IDE can cause configuration troubles, and it has cost me a lot of time dealing with issues related to vscode, webstorm, pnpm, and eslint. While I cannot easily remember specific problems, there are tons of them. I think anyone who uses these tools together has encountered some kind of problem, just like me.

**Comment by skimhugo** [comment]
I am tortured every day by configuration issues with monorepo, and I dare not easily switch to new configuration rules.

**Comment by eakl** [comment]
agree with @skimhugo, vscode + pnpm monorepo + eslint is just too much pain that I ended up not using eslint at all.

cf. my previous comment with an example repo: https://github.com/eslint/eslint/discussions/16960#discussioncomment-5369861.

For me, it doesn't make sense to locate eslint configs far from where they are used.
